src/casts/interpret-casts-common.rkt

```diff
@@ -1824,6 +1767,41 @@ TODO write unit tests
 (define monotonic-cast-close-code-specialization? : (Parameterof Boolean)
   (make-parameter #t))
 
+(define monotonic-blame (Quote "Monotonic references do not currently track blame"))
+
+(define (code-gen-mref-state-reduction
+         #:interp-cast interp-cast
+         #:greatest-lower-bound greatest-lower-bound)
+  (while$ (op$ mref-cast-queue-not-empty?)
+          (let*$ ([address (op$ mref-cast-queue-peek-address)]
+                  [t1 (Mbox-rtti-ref address)]                 
+                  [t2 (op$ mref-cast-queue-peek-type)]
+                  [t3 (app-code$ greatest-lower-bound t1 t2)])
+            (op$ mref-cast-queue-dequeue)
+            (when$ (not$ (op=? t1 t3))
+                   (let*$ ([vi (Mbox-val-ref address)]
+                           [cvi (interp-cast vi t1 t3 monotonic-blame (Quote #f))])
+                     (Mbox-rtti-set! address t3)
+                     (Mbox-val-set! address cvi))))))
+
+(define (code-gen-mvect-state-reduction
+         #:interp-cast interp-cast
+         #:greatest-lower-bound greatest-lower-bound)
+  (while$ (op$ mvect-cast-queue-not-empty?)
+          (let*$ ([address (op$ mvect-cast-queue-peek-address)]
+                  [t1 (Mvector-rtti-ref address)]                 
+                  [t2 (op$ mvect-cast-queue-peek-type)]
+                  [t3 (app-code$ greatest-lower-bound t1 t2)])
+            (op$ mvect-cast-queue-dequeue)
+            (when$ (not$ (op=? t1 t3))
+                   (let$ ([len (Mvector-length address)])
+                     (repeat$ (i ZERO-EXPR len) ()
+                       (let*$ ([vi (Mvector-val-ref address i 'no-check-bounds)]
+                               [cvi (interp-cast vi t1 t3 monotonic-blame (Quote #f))]
+                               [t4 (Mvector-rtti-ref address)])
+                         (Mvector-rtti-set! address t3)
```

Comment:
Is there a good reason for this being in the loop (this is loop invariant).  Want to move it before or after the loop, I am not sure which is better.

Reply:
Done.